src/librustc_middle/ty/query/plumbing.rs

```diff
@@ -234,18 +234,54 @@ macro_rules! hash_result {
 
 macro_rules! define_queries {
     (<$tcx:tt> $($category:tt {
-        $($(#[$attr:meta])* [$($modifiers:tt)*] fn $name:ident: $node:ident($K:ty) -> $V:ty,)*
+        $($(#[$attr:meta])* [$($modifiers:tt)*] fn $name:ident: $node:ident($($K:tt)*) -> $V:ty,)*
     },)*) => {
         define_queries_inner! { <$tcx>
-            $($( $(#[$attr])* category<$category> [$($modifiers)*] fn $name: $node($K) -> $V,)*)*
+            $($( $(#[$attr])* category<$category> [$($modifiers)*] fn $name: $node($($K)*) -> $V,)*)*
         }
     }
 }
 
+macro_rules! define_query_getter {
+    (TyCtxtAt<$tcx:tt>, $(#[$attr:meta])* $name:ident(DefId) -> $V:ty) => {
+        $(#[$attr])*
+        #[inline(always)]
+        pub fn $name(self, key: impl Into<DefId>) -> $V {
```

Comment:
Can you make a type macro and use *that* to pattern-match on just the key type?
I think we can just use `.into()` in all methods, as it will compile to a noop when the types are the same.

Reply:
That's a good idea. It won't work if we do want a monomorphic helper function, but maybe we don't.

Reply:
`get_query` is `#[inline(never)]` (I just checked) so you don't need this wrapper.